flow/scene_update_context.cc

```diff
@@ -4,124 +4,273 @@
-#include "apps/mozart/lib/skia/type_converters.h" // nogncheck
-#include "apps/mozart/services/composition/scenes.fidl.h" // nogncheck
+#include "flutter/flow/export_node.h"
 #include "flutter/flow/layers/layer.h"
+#include "flutter/flow/matrix_decomposition.h"
 #include "flutter/glue/trace_event.h"
 
 namespace flow {
 
-SceneUpdateContext::CurrentPaintTask::CurrentPaintTask()
-    : bounds(SkRect::MakeEmpty()) {}
-
-void SceneUpdateContext::CurrentPaintTask::Clear() {
-  bounds = SkRect::MakeEmpty();
-  layers.clear();
-}
-
-SceneUpdateContext::SceneUpdateContext(mozart::SceneUpdate* update,
+SceneUpdateContext::SceneUpdateContext(mozart::client::Session* session,
                                        SurfaceProducer* surface_producer)
-    : update_(update), surface_producer_(surface_producer) {}
+    : session_(session), surface_producer_(surface_producer) {
+  FTL_DCHECK(surface_producer_ != nullptr);
+}
 
 SceneUpdateContext::~SceneUpdateContext() = default;
 
-void SceneUpdateContext::AddLayerToCurrentPaintTask(Layer* layer) {
-  current_paint_task_.bounds.join(layer->paint_bounds());
-  current_paint_task_.layers.push_back(layer);
-}
+void SceneUpdateContext::AddChildScene(ExportNode* export_node,
+                                       SkPoint offset,
+                                       float device_pixel_ratio,
+                                       bool hit_testable) {
+  FTL_DCHECK(top_entity_);
 
-void SceneUpdateContext::FinalizeCurrentPaintTaskIfNeeded(
-    mozart::Node* container,
-    const SkMatrix& ctm) {
-  if (mozart::NodePtr node = FinalizeCurrentPaintTask(ctm))
-    AddChildNode(container, std::move(node));
+  export_node->Bind(*this, top_entity_->entity_node(), offset,
+                    1.f / device_pixel_ratio, hit_testable);
 }
 
-mozart::NodePtr SceneUpdateContext::FinalizeCurrentPaintTask(
-    const SkMatrix& ctm) {
-  if (current_paint_task_.layers.empty())
-    return nullptr;
+void SceneUpdateContext::CreateFrame(mozart::client::EntityNode& entity_node,
+                                     const SkRRect& rrect,
+                                     SkColor color,
+                                     SkScalar scale_x,
+                                     SkScalar scale_y,
+                                     const SkRect& paint_bounds,
+                                     std::vector<Layer*> paint_layers) {
+  // Frames always clip their children.
+  entity_node.SetClip(0u, true /* clip to self */);
 
-  const SkRect& bounds = current_paint_task_.bounds;
+  // We don't need a shape if the frame is zero size.
+  if (rrect.isEmpty())
+    return;
 
-  SkScalar scaleX = ctm.getScaleX();
-  SkScalar scaleY = ctm.getScaleY();
+  // Add a part which represents the frame's geometry for clipping purposes
+  // and possibly for its texture.
+  // TODO(MZ-137): Need to be able to express the radii as vectors.
+  // TODO(MZ-138): Need to be able to specify an origin.
```

Comment:
I think the transform is sufficient to set the origin (and we already apply it).  So let's remove this TODO here.